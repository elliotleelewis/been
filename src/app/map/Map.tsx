import { centerOfMass, featureCollection } from '@turf/turf';
import { type FC, ReactNode, useEffect, useMemo, useRef } from 'react';
import {
	Layer,
	MapRef,
	NavigationControl,
	Map as ReactMapGL,
	Source,
} from 'react-map-gl';

import { useCountries } from '../contexts/CountriesContext';
import { useMatchMedia } from '../hooks/useMatchMedia';

const apiKeyMapbox = import.meta.env['VITE_API_KEY_MAPBOX'] as
	| string
	| undefined;
const darkThemeUrl = 'mapbox://styles/mapbox/dark-v11';
const lightThemeUrl = 'mapbox://styles/mapbox/light-v11';

const minZoom = 1.8;

interface Props {
	header: ReactNode;
}

export const Map: FC<Props> = () => {
	const mapRef = useRef<MapRef>(null);
	const prefersDark = useMatchMedia('(prefers-color-scheme: dark)');

	const { countries, focus } = useCountries();

	const selectedCountries = useMemo(() => {
		return countries.filter((c) => c.selected).map((c) => c.iso3166);
	}, [countries]);

	useEffect(() => {
		if (!focus || !mapRef.current) {
			return;
		}
		const { current: map } = mapRef;
		const features = map.querySourceFeatures('countries', {
			sourceLayer: 'country_boundaries',
			filter: ['==', ['get', 'iso_3166_1'], focus],
		});
		if (features.length > 0) {
			const { geometry } = centerOfMass(featureCollection(features));
			const [lat = 0, lon = 0] = geometry.coordinates;
			map.flyTo({ center: [lat, lon], zoom: minZoom });
		}
	}, [focus, mapRef]);

	if (!apiKeyMapbox) {
		return (
			<div className="flex size-full items-center justify-center bg-primary text-white dark:bg-zinc-950 dark:text-primary">
				<div>Whoops, something has gone wrong.</div>
			</div>
		);
	}

	return (
		<ReactMapGL
			mapboxAccessToken={apiKeyMapbox}
			mapStyle={prefersDark ? darkThemeUrl : lightThemeUrl}
			antialias
			attributionControl={false}
			minZoom={minZoom}
			ref={mapRef}
		>
			<NavigationControl />
			<Source
				id="countries"
				type="vector"
				url="mapbox://mapbox.country-boundaries-v1"
			/>
			<Layer
				id="visited-countries"
				type="fill"
				source="countries"
				source-layer="country_boundaries"
				beforeId="national-park"
				filter={[
					'in',
					['get', 'iso_3166_1'],
					['literal', selectedCountries],
				]}
				paint={{ 'fill-color': '#fd7e14', 'fill-opacity': 0.6 }}
			/>
			<Layer
				id="3d-buildings"
				type="fill-extrusion"
				source="composite"
				source-layer="building"
				minzoom={15}
				filter={['==', 'extrude', 'true']}
				paint={{
					'fill-extrusion-color': [
						'case',
						[
							'in',
							['get', 'iso_3166_1'],
							['literal', selectedCountries],
						],
						'#fd7e14',
						'#fd7e14',
					],
					'fill-extrusion-height': [
						'interpolate',
						['linear'],
						['zoom'],
						15,
						0,
						15.05,
						['get', 'height'],
					],
					'fill-extrusion-base': [
						'interpolate',
						['linear'],
						['zoom'],
						15,
						0,
						15.05,
						['get', 'min_height'],
					],
					'fill-extrusion-opacity': 0.6,
				}}
			/>
		</ReactMapGL>
	);
};
