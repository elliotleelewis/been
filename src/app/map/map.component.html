<mgl-map
	logoPosition="bottom-right"
	[attributionControl]="false"
	[minZoom]="minZoom"
	[style]="'mapbox://styles/elliotleelewis/ckj1w9ro12g8h19mqhvsgcbkg'"
	[zoom]="[minZoom]"
>
	<mgl-vector-source
		id="countries"
		url="mapbox://mapbox.country-boundaries-v1"
	></mgl-vector-source>
	<mgl-layer
		id="visited-countries"
		type="fill"
		source="countries"
		sourceLayer="country_boundaries"
		[filter]="filter"
		[paint]="{ 'fill-color': '#fd7e14', 'fill-opacity': 0.7 }"
	></mgl-layer>
</mgl-map>
