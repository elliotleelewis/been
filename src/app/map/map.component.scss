:host {
	mgl-map {
		width: 100%;
		height: 100%;

		::ng-deep {
			canvas {
				outline: none;
			}
		}
	}
}
