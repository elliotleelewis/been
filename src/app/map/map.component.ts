import {
	ChangeDetectionStrategy,
	Component,
	HostBinding,
	Inject,
	type OnDestroy,
	type OnInit,
} from '@angular/core';
import { type Map } from 'mapbox-gl';
import { type Observable, distinctUntilChanged, fromEvent, map } from 'rxjs';
import { SubSink } from 'subsink';

import {
	type Coordinate,
	type NestedCoordinates,
	getCoordsCenter,
} from '../helpers';
import { CountriesService } from '../services/countries.service';

@Component({
	selector: 'app-map',
	templateUrl: './map.component.html',
	styleUrls: ['./map.component.scss'],
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MapComponent implements OnInit, OnDestroy {
	@HostBinding('class')
	class = 'flex flex-col';

	readonly darkThemeUrl = 'mapbox://styles/mapbox/dark-v11';
	readonly lightThemeUrl = 'mapbox://styles/mapbox/light-v11';
	readonly prefersDark = window.matchMedia('(prefers-color-scheme: dark)')
		.matches;

	map?: Map;
	center?: Coordinate;
	minZoom = 1.8;

	private _subs = new SubSink();

	constructor(
		@Inject(CountriesService) private countriesService: CountriesService,
	) {}

	static get prefersDark$(): Observable<boolean> {
		return fromEvent<MediaQueryListEvent>(
			window.matchMedia('(prefers-color-scheme: dark)'),
			'change',
		).pipe(
			map((event) => event.matches),
			distinctUntilChanged(),
		);
	}

	get countriesSelected$(): Observable<string[]> {
		return this.countriesService.countries$.pipe(
			map((countries) =>
				countries.filter((c) => c.selected).map((c) => c.iso3166),
			),
		);
	}

	ngOnInit(): void {
		this._subs.sink = this.countriesService.focus$.subscribe((country) => {
			const query = this.map?.querySourceFeatures('countries', {
				sourceLayer: 'country_boundaries',
				filter: ['==', ['get', 'iso_3166_1'], country],
			});
			const feature = query?.[0];
			if (feature) {
				const { coordinates } = feature.geometry as {
					coordinates?: NestedCoordinates;
				};
				if (coordinates) {
					this.center = getCoordsCenter(coordinates);
				}
			}
		});

		this._subs.sink = MapComponent.prefersDark$.subscribe((prefersDark) => {
			this.map?.setStyle(
				prefersDark ? this.darkThemeUrl : this.lightThemeUrl,
			);
		});
	}

	ngOnDestroy(): void {
		this._subs.unsubscribe();
	}

	onMapLoad(map: Map): void {
		this.map = map;
	}
}
