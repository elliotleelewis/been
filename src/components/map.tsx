import { centerOfMass, featureCollection } from '@turf/turf';
import { type FillExtrusionPaint, type FillPaint } from 'mapbox-gl';
import {
	type ReactNode,
	forwardRef,
	memo,
	useEffect,
	useImperativeHandle,
	useMemo,
	useRef,
} from 'react';
import {
	Layer,
	type MapRef,
	NavigationControl,
	Map as ReactMapGL,
	Source,
} from 'react-map-gl';

import { useCountries } from '../contexts/countries-context';
import { useMatchMedia } from '../hooks/use-match-media';
import { MapboxLayerKeys, MapboxSourceKeys } from '../models/enums';
import { type ForwardedRefFunction } from '../types/utils';

const apiKeyMapbox = import.meta.env['VITE_API_KEY_MAPBOX'] as
	| string
	| undefined;
const testMode = import.meta.env.MODE === 'test';
const darkThemeUrl = 'mapbox://styles/mapbox/dark-v11';
const lightThemeUrl = 'mapbox://styles/mapbox/light-v11';

const minZoom = 1.8;

export interface MapForwardedRef {
	isSourceLoaded: ForwardedRefFunction<MapRef['isSourceLoaded']>;
	querySourceFeatures: ForwardedRefFunction<MapRef['querySourceFeatures']>;
}

interface Props {
	header: ReactNode;
}

export const Map = memo(
	forwardRef<MapForwardedRef, Props>(({ header }, ref) => {
		const internalRef = useRef<MapRef>(null);

		const prefersDark = useMatchMedia('(prefers-color-scheme: dark)');

		const { countries, focus } = useCountries();

		const selectedCountries = useMemo(() => {
			return countries.filter((c) => c.selected).map((c) => c.iso3166);
		}, [countries]);

		useImperativeHandle(
			/* eslint-disable react-compiler/react-compiler -- This is a bug with the ESLint plugin, this is correct behaviour according to the React docs (see here: https://react.dev/reference/react/useImperativeHandle#exposing-a-custom-ref-handle-to-the-parent-component). */
			ref,
			() => ({
				isSourceLoaded: (
					...params: Parameters<MapRef['isSourceLoaded']>
				) => {
					return internalRef.current?.isSourceLoaded(...params);
				},
				querySourceFeatures: (
					...params: Parameters<MapRef['querySourceFeatures']>
				) => {
					return internalRef.current?.querySourceFeatures(...params);
				},
			}),
			/* eslint-enable react-compiler/react-compiler */
			[],
		);

		useEffect(() => {
			if (!focus || !internalRef.current) {
				return;
			}
			const { current: map } = internalRef;
			const features = map.querySourceFeatures(
				MapboxSourceKeys.countries,
				{
					sourceLayer: 'country_boundaries',
					filter: ['==', ['get', 'iso_3166_1'], focus],
				},
			);
			if (features.length > 0) {
				const { geometry } = centerOfMass(featureCollection(features));
				const [lat = 0, lon = 0] = geometry.coordinates;
				map.flyTo({ center: [lat, lon], zoom: minZoom });
			}
		}, [focus, internalRef]);

		const beenFilter = useMemo(
			() => ['in', ['get', 'iso_3166_1'], ['literal', selectedCountries]],
			[selectedCountries],
		);
		const beenPaint: FillPaint = useMemo(
			() => ({
				/* eslint-disable @typescript-eslint/naming-convention -- mapbox-gl has specific property names we must use */
				'fill-color': '#fd7e14',
				'fill-opacity': 0.6,
				/* eslint-enable @typescript-eslint/naming-convention */
			}),
			[],
		);

		const buildingsFilter = useMemo(() => ['==', 'extrude', 'true'], []);
		const buildingsPaint: FillExtrusionPaint = useMemo(
			() => ({
				/* eslint-disable @typescript-eslint/naming-convention -- mapbox-gl has specific property names we must use */
				'fill-extrusion-color': [
					'case',
					[
						'in',
						['get', 'iso_3166_1'],
						['literal', selectedCountries],
					],
					'#fd7e14',
					'#fd7e14',
				],
				'fill-extrusion-height': [
					'interpolate',
					['linear'],
					['zoom'],
					15,
					0,
					15.05,
					['get', 'height'],
				],
				'fill-extrusion-base': [
					'interpolate',
					['linear'],
					['zoom'],
					15,
					0,
					15.05,
					['get', 'min_height'],
				],
				'fill-extrusion-opacity': 0.6,
				/* eslint-enable @typescript-eslint/naming-convention */
			}),
			[selectedCountries],
		);

		return (
			<>
				{header}
				<ReactMapGL
					mapboxAccessToken={apiKeyMapbox ?? ''}
					mapStyle={prefersDark ? darkThemeUrl : lightThemeUrl}
					antialias
					attributionControl={false}
					logoPosition="bottom-right"
					minZoom={minZoom}
					ref={internalRef}
					testMode={testMode}
				>
					<NavigationControl />
					<Source
						id={MapboxSourceKeys.countries}
						type="vector"
						url="mapbox://mapbox.country-boundaries-v1"
					/>
					<Layer
						id={MapboxLayerKeys.been}
						type="fill"
						source={MapboxSourceKeys.countries}
						source-layer="country_boundaries"
						beforeId="national-park"
						filter={beenFilter}
						paint={beenPaint}
					/>
					<Layer
						id={MapboxLayerKeys.buildings}
						type="fill-extrusion"
						source="composite"
						source-layer="building"
						minzoom={15}
						filter={buildingsFilter}
						paint={buildingsPaint}
					/>
				</ReactMapGL>
			</>
		);
	}),
);
Map.displayName = 'Map';
