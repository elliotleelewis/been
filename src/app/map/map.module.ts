import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { MAPBOX_API_KEY, NgxMapboxGLModule } from 'ngx-mapbox-gl';

import { ENVIRONMENT } from '../../environments/environment';

import { MapComponent } from './map.component';

@NgModule({
	declarations: [MapComponent],
	imports: [BrowserModule, NgxMapboxGLModule],
	exports: [MapComponent],
	providers: [
		{
			provide: MAPBOX_API_KEY,
			useValue: ENVIRONMENT.apiKeyMapbox,
		},
	],
})
export class MapModule {}
