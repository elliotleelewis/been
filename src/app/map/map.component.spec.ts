import type { ComponentFixture } from '@angular/core/testing';
import { TestBed } from '@angular/core/testing';
import { MockModule, MockProvider } from 'ng-mocks';
import { NgxMapboxGLModule } from 'ngx-mapbox-gl';
import { NEVER, of } from 'rxjs';

import { CountriesService } from '../services/countries.service';

import { MapComponent } from './map.component';

describe('MapComponent', () => {
	let component: MapComponent;
	let fixture: ComponentFixture<MapComponent>;

	beforeEach(async () => {
		await TestBed.configureTestingModule({
			declarations: [MapComponent],
			imports: [MockModule(NgxMapboxGLModule)],
			providers: [
				MockProvider(CountriesService, {
					countries$: of([]),
					focus$: NEVER,
				}),
			],
		}).compileComponents();
	});

	beforeEach(() => {
		fixture = TestBed.createComponent(MapComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('should render', () => {
		expect(fixture).toMatchSnapshot();
	});
});
