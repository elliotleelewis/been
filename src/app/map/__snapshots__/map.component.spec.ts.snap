// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapComponent should render 1`] = `
<app-map
  __ngContext__={[Function LRootView]}
  _subs={[Function SubSink]}
  countriesService={[Function CountriesService]}
  filter={[Function Array]}
  minZoom={[Function Number]}
>
  <mgl-map
    logoposition="bottom-right"
  >
    <mgl-vector-source
      id="countries"
      url="mapbox://mapbox.country-boundaries-v1"
    />
    <mgl-layer
      id="visited-countries"
      source="countries"
      sourcelayer="country_boundaries"
      type="fill"
    />
  </mgl-map>
</app-map>
`;
